File: LR2/src/main/kotlin/types/users/IntegerUserType.kt
package types.users

import comparator.Comparator
import comparator.IntegerComparator
import types.IntegerType
import java.io.IOException
import java.io.InputStreamReader
import java.util.*

class IntegerUserType : UserType {
    private val MAX = 1000
    private val MIN = -1000
    override fun typeName(): String {
        return "Integer"
    }

    @Throws(InstantiationException::class)
    override fun create(): Any {
        val random = Random()
        val value = random.nextInt(MAX - MIN) + MIN
        return IntegerType(value)
    }

    @Throws(InstantiationException::class)
    override fun clone(`object`: Any): Any {
        return IntegerType((`object` as IntegerType).getIntegerValue())
    }

    override fun readValue(`in`: InputStreamReader?): Any {
        return try {
            `in`!!.read()
        } catch (e: IOException) {
            throw RuntimeException(e)
        }
    }

    @Throws(InstantiationException::class)
    override fun parseValue(ss: String): Any? {
        return try {
            IntegerType(ss.toInt())
        } catch (e: NumberFormatException) {
            null
        }
    }

    override val typeComparator: Comparator
        get() = IntegerComparator()

    override fun toString(`object`: Any): String {
        return `object`.toString()
    }
}

File: LR1/src/main/scala/types/users/IntegerUserType.scala
package types.users

import comparator.IntegerComparator
import java.io.IOException
import java.io.InputStreamReader
import java.util.Random
import types.IntegerType

class IntegerUserType extends UserType {
  final private val MAX = 1000
  final private val MIN = -1000

  override def typeName = "Integer"

  @throws[InstantiationException]
  override def create: AnyRef = {
    val random = new Random
    val value = random.nextInt(MAX - MIN) + MIN
    new IntegerType(value)
  }

  @throws[InstantiationException]
  override def clone(`object`: AnyRef) = new IntegerType(`object`.asInstanceOf[IntegerType].getIntegerValue)

  override def readValue(in: InputStreamReader): Any = try in.read
  catch {
    case e: IOException =>
      throw new RuntimeException(e)
  }

  @throws[InstantiationException]
  override def parseValue(ss: String): AnyRef = try new IntegerType(ss.toInt)
  catch {
    case e: NumberFormatException =>
      null
  }

  override def getTypeComparator = new IntegerComparator

  override def toString(`object`: AnyRef): String = `object`.toString
}